Content.Server/Sports/PitchingMachineSystem.cs

```diff
@@ -0,0 +1,119 @@
+using System.Linq;
+using Content.Server.Popups;
+using Content.Server.Sports.Components;
+using Content.Server.Storage.Components;
+using Content.Server.Storage.EntitySystems;
+using Content.Shared.Throwing;
+using Content.Shared.Verbs;
+using Robust.Shared.Audio;
+using Robust.Shared.Physics;
+using Robust.Shared.Player;
+using Robust.Shared.Random;
+
+namespace Content.Server.Sports
+{
+    public sealed class PitchingMachineSystem : EntitySystem
+    {
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
+        [Dependency] private readonly ThrowingSystem _throwingSystem = default!;
+        [Dependency] private readonly IRobustRandom _robustRandom = default!;
+        [Dependency] private readonly SharedAudioSystem _audioSystem = default!;
+        [Dependency] private readonly StorageSystem _storageSystem = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<PitchingMachineComponent, GetVerbsEvent<Verb>>(AddEjectVerb);
+            SubscribeLocalEvent<PitchingMachineComponent, GetVerbsEvent<AlternativeVerb>>(AddPowerVerb);
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+            foreach (var ballLauncher in EntityQuery<PitchingMachineComponent>())
+            {
+                ballLauncher.AccumulatedFrametime += frameTime;
+
+                if (ballLauncher.AccumulatedFrametime < ballLauncher.CurrentLauncherCooldown)
+                    continue;
+
+                ballLauncher.AccumulatedFrametime -= ballLauncher.CurrentLauncherCooldown;
+                ballLauncher.CurrentLauncherCooldown = ballLauncher.ShootCooldown;
+
+                if (ballLauncher.IsOn)
+                    Fire(ballLauncher.Owner);
+            }
+        }
+
+        public void TogglePower(EntityUid uid, PitchingMachineComponent component)
+        {
+            if (!component.IsOn)
+            {
+                component.IsOn = true;
+                _popupSystem.PopupEntity(Loc.GetString("comp-emitter-turned-on", ("target", component.Owner)), component.Owner, Filter.Pvs(component.Owner));
+            }
+            else
+            {
+                component.IsOn = false;
+                _popupSystem.PopupEntity(Loc.GetString("comp-emitter-turned-off", ("target", component.Owner)), component.Owner, Filter.Pvs(component.Owner));
+            }
+        }
+
+        private void Fire(EntityUid uid)
```

Comment:
I think you can make many things in the pitcher use `GunSystem`/`GunComponent` and `BasicEntityAmmoProviderComponent` so take a look at those (you can check `WeaponStaffBase`, `WeaponWandBase`and `WeaponProtoKineticAcceleratorBase` for examples)

Reply:
I wouldn't know how to do this, the pitching machine isn't supposed to take just baseballs. It uses any item in it's storage as ammo and simulates a throw. I only know of pneumatic cannons doing this and they don't use GunSystem. I also don't know of any examples of an unmanned turret using GunSystem.

Reply:
Pneumatic cannon is oldcode that needs updating to use the new guncode because it's not predicted and doesn't benefit from updates to guns.